addons/auth_signup/models/res_partner.py

```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import random
+import werkzeug
+
+from datetime import datetime, timedelta
+from urlparse import urljoin
+
+from odoo import api, fields, models, _
+
+
+class SignupError(Exception):
+    pass
+
+def random_token():
+    # the token has an entropy of about 120 bits (6 bits/char * 20 chars)
+    chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
+    return ''.join(random.SystemRandom().choice(chars) for i in xrange(20))
+
+def now(**kwargs):
+    dt = datetime.now() + timedelta(**kwargs)
+    return fields.Datetime.to_string(dt)
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    signup_token = fields.Char(copy=False)
+    signup_type = fields.Char(string='Signup Token Type', copy=False)
+    signup_expiration = fields.Datetime(copy=False)
+    signup_valid = fields.Boolean(compute='_compute_signup_valid', string='Signup Token is Valid')
+    signup_url = fields.Char(compute='_compute_signup_url', string='Signup URL')
+
+    @api.multi
+    def _compute_signup_valid(self):
+        for partner in self:
+            partner.signup_valid = bool(partner.signup_token) and \
+            (not partner.signup_expiration or now() <= partner.signup_expiration)
```

Comment:
in previous code, the call to `now()` was set on a variable. This one is quite fast but we usually avoid to execute a method for each partner (e.g. if we compute the value on 1000 partners) if the result is going to be the same for each anyway.